store/mongo/control_reader.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2021 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package mongo
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/binary"
+	"io"
+
+	"go.mongodb.org/mongo-driver/mongo"
+
+	"github.com/mendersoftware/deviceconnect/app"
+	"github.com/mendersoftware/deviceconnect/model"
+)
+
+const (
+	controlReadBufferSize = 4096
+)
+
+var (
+	noopReader = &ControlMessageReader{}
+)
+
+type ControlMessageReader struct {
+	currentOffset int
+	output        []byte
+	outputLength  int
+	buffer        bytes.Buffer
+	c             *mongo.Cursor
+	gzipReader    *gzip.Reader
+}
+
+func NewControlMessageReader(ctx context.Context, c *mongo.Cursor) *ControlMessageReader {
+	reader := &ControlMessageReader{
+		currentOffset: 0,
+		output:        make([]byte, controlReadBufferSize),
+		buffer:        bytes.Buffer{},
+		c:             c,
+		outputLength:  0,
+	}
+
+	c.Next(ctx)
```

Comment:
Return value is not checked; what if the query does not match any documents?

Reply:
the Decode in the following lines would fail and the noopReader would be returned, which will return nil from the first Pop and end the calls to this reader. I will just add the check you mentioned and return noop earlier.